File: common/src/main/java/com/vladte/devhack/common/service/domain/global/TagService.java
package com.vladte.devhack.common.service.domain.global;

import com.vladte.devhack.common.service.domain.CrudService;
import com.vladte.devhack.entities.global.Tag;
import com.vladte.devhack.entities.user.User;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

/**
 * Service interface for Tag entity operations with hierarchical support.
 */
public interface TagService extends CrudService<Tag, UUID> {
    /**
     * Find a tag by name.
     *
     * @param name the tag name
     * @return an Optional containing the tag, or empty if not found
     */
    Optional<Tag> findTagByName(String name);

    /**
     * Find a tag by slug.
     *
     * @param slug the tag slug
     * @return an Optional containing the tag, or empty if not found
     */
    Optional<Tag> findTagBySlug(String slug);

    /**
     * Create a new tag with proper path generation.
     *
     * @param name   the tag name
     * @param parent the parent tag (null for root tags)
     * @return the created tag
     */
    Tag createTag(String name, Tag parent);

    /**
     * Move a tag to a new parent, updating all subtree paths.
     *
     * @param tag       the tag to move
     * @param newParent the new parent (null to make it a root tag)
     * @return the updated tag
     */
    Tag moveTag(Tag tag, Tag newParent);

    /**
     * Delete a tag with cascade or orphan handling policy.
     *
     * @param tag           the tag to delete
     * @param cascadeDelete if true, delete all children; if false, move children to root
     * @return true if deletion was successful
     */
    boolean deleteTag(Tag tag, boolean cascadeDelete);

    /**
     * Find all root tags (tags with no parent).
     *
     * @return list of root tags
     */
    List<Tag> findRootTags();

    /**
     * Find all direct children of a tag.
     *
     * @param parent the parent tag
     * @return list of direct children
     */
    List<Tag> findChildren(Tag parent);

    /**
     * Find all descendants of a tag (entire subtree).
     *
     * @param parent the parent tag
     * @return list of all descendants
     */
    List<Tag> findDescendants(Tag parent);

    /**
     * Find all ancestors of a tag.
     *
     * @param tag the tag
     * @return list of ancestors from root to immediate parent
     */
    List<Tag> findAncestors(Tag tag);

    /**
     * Find subtree limited by depth.
     *
     * @param parent the parent tag
     * @param depth  maximum depth to include
     * @return list of tags in subtree within depth limit
     */
    List<Tag> findSubtree(Tag parent, int depth);

    /**
     * Find siblings of a tag (tags with same parent).
     *
     * @param tag the tag
     * @return list of sibling tags
     */
    List<Tag> findSiblings(Tag tag);

    /**
     * Find tags at a specific depth level.
     *
     * @param depth the depth level (0 for root)
     * @return list of tags at the specified depth
     */
    List<Tag> findTagsByDepth(int depth);

    /**
     * Validate that moving a tag won't create a cycle.
     *
     * @param tag       the tag to move
     * @param newParent the proposed new parent
     * @return true if the move is valid (no cycle)
     */
    boolean validateMove(Tag tag, Tag newParent);

    /**
     * Calculate progress for a tag based on the user's answered questions.
     *
     * @param tag  the tag to calculate progress for
     * @param user the user to calculate progress for
     * @return the tag with updated progress information
     */
    Tag calculateProgress(Tag tag, User user);

    /**
     * Calculate progress for all tags based on the user's answered questions.
     *
     * @param tags the list of tags to calculate progress for
     * @param user the user to calculate progress for
     * @return the list of tags with updated progress information
     */
    List<Tag> calculateProgressForAll(List<Tag> tags, User user);

    /**
     * Count tags used by a specific user in their questions.
     *
     * @param user the user
     * @return the count of tags used by the user
     */
    int countTagsByUser(User user);

    /**
     * Count all tags in the system.
     *
     * @return the total count of tags
     */
    int countAllTags();
}

File: angular-ui/src/app/services/global/tag.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Observable} from 'rxjs';
import {
    TagCreateRequest,
    TagDTO,
    TagHierarchyResponse,
    TagMoveRequest,
    TagTreeNode,
    TagValidationResponse
} from '../../models/global/tag.model';
import {BaseService} from '../base.service';

/**
 * Service for managing tags.
 * Extends BaseService to inherit common CRUD operations and follows OOP principles.
 */
@Injectable({
    providedIn: 'root'
})
export class TagService extends BaseService<TagDTO> {
    protected baseUrl = '/api/tags';

    constructor(http: HttpClient) {
        super(http);
    }

    /**
     * Get tag by name
     * @param name - Tag name
     * @returns Observable tag
     */
    getByName(name: string): Observable<TagDTO> {
        return this.http.get<TagDTO>(`${this.baseUrl}/by-name/${name}`);
    }

    /**
     * Search tags by query
     * @param query - Search query
     * @returns Observable array of tags
     */
    override search(searchParams: { [key: string]: string | undefined }): Observable<TagDTO[]>;
    override search(query: string): Observable<TagDTO[]>;
    override search(queryOrParams: string | { [key: string]: string | undefined }): Observable<TagDTO[]> {
        if (typeof queryOrParams === 'string') {
            return super.search({query: queryOrParams});
        }
        return super.search(queryOrParams);
    }

    /**
     * Get popular tags with optional limit
     * @param limit - Optional limit for number of tags
     * @returns Observable array of popular tags
     */
    getPopular(limit?: number): Observable<TagDTO[]> {
        return this.getWithParams('popular', {limit});
    }

    /**
     * Get tags with their question count
     * @returns Observable array of tags with question count
     */
    getTagsWithQuestionCount(): Observable<any[]> {
        return this.http.get<any[]>(`${this.baseUrl}/with-question-count`);
    }

    // ========== Hierarchical Methods ==========

    /**
     * Create a new tag with optional parent
     * @param request - Tag creation request with parent information
     * @returns Observable created tag
     */
    createWithParent(request: TagCreateRequest): Observable<TagDTO> {
        return this.http.post<TagDTO>(`${this.baseUrl}/hierarchy`, request);
    }

    /**
     * Move a tag to a new parent
     * @param request - Tag move request
     * @returns Observable updated tag
     */
    moveTag(request: TagMoveRequest): Observable<TagDTO> {
        return this.http.put<TagDTO>(`${this.baseUrl}/hierarchy/move`, request);
    }

    /**
     * Delete a tag with cascade option
     * @param id - Tag ID
     * @param cascade - Whether to cascade delete children
     * @returns Observable void
     */
    deleteWithCascade(id: string, cascade: boolean = false): Observable<void> {
        return this.http.delete<void>(`${this.baseUrl}/hierarchy/${id}?cascade=${cascade}`);
    }

    /**
     * Get all root tags (tags with no parent)
     * @returns Observable array of root tags
     */
    getRootTags(): Observable<TagDTO[]> {
        return this.http.get<TagDTO[]>(`${this.baseUrl}/hierarchy/roots`);
    }

    /**
     * Get direct children of a tag
     * @param parentId - Parent tag ID
     * @returns Observable array of child tags
     */
    getChildren(parentId: string): Observable<TagDTO[]> {
        return this.http.get<TagDTO[]>(`${this.baseUrl}/hierarchy/${parentId}/children`);
    }

    /**
     * Get all descendants of a tag (entire subtree)
     * @param parentId - Parent tag ID
     * @returns Observable array of descendant tags
     */
    getDescendants(parentId: string): Observable<TagDTO[]> {
        return this.http.get<TagDTO[]>(`${this.baseUrl}/hierarchy/${parentId}/descendants`);
    }

    /**
     * Get all ancestors of a tag
     * @param tagId - Tag ID
     * @returns Observable array of ancestor tags
     */
    getAncestors(tagId: string): Observable<TagDTO[]> {
        return this.http.get<TagDTO[]>(`${this.baseUrl}/hierarchy/${tagId}/ancestors`);
    }

    /**
     * Get subtree with limited depth
     * @param parentId - Parent tag ID
     * @param depth - Maximum depth to include
     * @returns Observable array of tags in subtree
     */
    getSubtree(parentId: string, depth: number): Observable<TagDTO[]> {
        return this.http.get<TagDTO[]>(`${this.baseUrl}/hierarchy/${parentId}/subtree?depth=${depth}`);
    }

    /**
     * Get siblings of a tag
     * @param tagId - Tag ID
     * @returns Observable array of sibling tags
     */
    getSiblings(tagId: string): Observable<TagDTO[]> {
        return this.http.get<TagDTO[]>(`${this.baseUrl}/hierarchy/${tagId}/siblings`);
    }

    /**
     * Get tags at specific depth level
     * @param depth - Depth level (0 for root)
     * @returns Observable array of tags at depth
     */
    getTagsByDepth(depth: number): Observable<TagDTO[]> {
        return this.http.get<TagDTO[]>(`${this.baseUrl}/hierarchy/depth/${depth}`);
    }

    /**
     * Validate if a tag move is valid (no cycles)
     * @param tagId - Tag ID to move
     * @param newParentId - New parent ID
     * @returns Observable validation response
     */
    validateMove(tagId: string, newParentId?: string): Observable<TagValidationResponse> {
        const params = newParentId ? `?newParentId=${newParentId}` : '';
        return this.http.get<TagValidationResponse>(`${this.baseUrl}/hierarchy/${tagId}/validate-move${params}`);
    }

    /**
     * Get complete tag hierarchy
     * @returns Observable hierarchy response with root tags and all tags
     */
    getHierarchy(): Observable<TagHierarchyResponse> {
        return this.http.get<TagHierarchyResponse>(`${this.baseUrl}/hierarchy`);
    }

    /**
     * Build tree structure from flat tag list
     * @param tags - Flat array of tags
     * @returns Tree structure
     */
    buildTree(tags: TagDTO[]): TagTreeNode[] {
        const tagMap = new Map<string, TagTreeNode>();
        const rootNodes: TagTreeNode[] = [];

        // Create nodes for all tags
        tags.forEach(tag => {
            if (tag.id) {
                tagMap.set(tag.id, {
                    tag,
                    children: [],
                    expanded: false,
                    level: tag.depth || 0
                });
            }
        });

        // Build tree structure
        tags.forEach(tag => {
            if (tag.id) {
                const node = tagMap.get(tag.id)!;
                if (tag.parent && tag.parent.id) {
                    const parentNode = tagMap.get(tag.parent.id);
                    if (parentNode) {
                        parentNode.children.push(node);
                    }
                } else {
                    rootNodes.push(node);
                }
            }
        });

        return rootNodes;
    }

    /**
     * Flatten tree structure to array
     * @param nodes - Tree nodes
     * @param includeCollapsed - Whether to include collapsed children
     * @returns Flat array of tree nodes
     */
    flattenTree(nodes: TagTreeNode[], includeCollapsed: boolean = true): TagTreeNode[] {
        const result: TagTreeNode[] = [];

        const traverse = (nodeList: TagTreeNode[]) => {
            nodeList.forEach(node => {
                result.push(node);
                if (node.expanded || includeCollapsed) {
                    traverse(node.children);
                }
            });
        };

        traverse(nodes);
        return result;
    }
}
